### PR TITLE
[CALCITE-3649] Hints should be propagated correctly in planner rules …

### PR DESCRIPTION
**Pull request by danny0405**
…if original node is transformed to different kind

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/test/SqlHintsConverterTest.java [hunk 1]
```diff
@@ -387,6 +389,25 @@ protected DiffRepository getDiffRepos() {
...
+        + "ename, avg(sal)\n"
+        + "from emp group by ename";
+    final RelNode rel = tester.convertSqlToRel(sql).rel;
+    final RelHint hint = RelHint.of(
+        Collections.emptyList(),
+        "AGG_STRATEGY",
+        Collections.singletonList("TWO_PHASE"));
+    // Validate Hep planner.
+    HepProgram program = new HepProgramBuilder()
+        .addRuleInstance(AggregateReduceFunctionsRule.INSTANCE)
```
Comment (amaliujia) [inline core/src/test/java/org/apache/calcite/test/SqlHintsConverterTest.java]: Maybe write a mock rule (similar to `MockJoin`) to simulate the `Project + Aggregate` split that you want to test?


I had a hard time to understand why `AggregateReduceFunctionsRule` will trigger what expects be tested?

File: core/src/main/java/org/apache/calcite/plan/RelOptUtil.java [hunk 2]
```diff
@@ -396,15 +396,60 @@ public static void verifyTypeEquivalence(
    * Copy the {@link org.apache.calcite.rel.hint.RelHint}s from {@code originalRel}
    * to {@code newRel} if both of them are {@link Hintable}.
    *
-   * <p>The hints are filtered by the hint strategies when attaching to the {@code newRel}.
+   * <p>The two relational expressions are assumed as semantically equivalent,
+   * that means the hints should be attached to the relational expression
+   * that expect to have them.
```
Comment (amaliujia) [inline core/src/main/java/org/apache/calcite/plan/RelOptUtil.java]: typo? maybe `expects or is expected`?

File: core/src/test/java/org/apache/calcite/test/SqlHintsConverterTest.java (hunk 1 above)
Comment (danny0405) [reply core/src/test/java/org/apache/calcite/test/SqlHintsConverterTest.java]: The aim is to test the real rules can propagate the hints successfully. If you also find any bad case, welcome to contribute more.

File: core/src/main/java/org/apache/calcite/plan/RelOptUtil.java (hunk 2 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/plan/RelOptUtil.java]: Yeah, thanks.

**Comment by cshuo** [comment]
LGTM~

**Comment by cshuo** [comment]
BTW, I noticed that the docs in `hint/package-info.java` should be updated 😄.

File: core/src/main/java/org/apache/calcite/plan/RelOptUtil.java [hunk 3]
```diff
@@ -396,15 +396,60 @@ public static void verifyTypeEquivalence(
...
-   * <p>The hints are filtered by the hint strategies when attaching to the {@code newRel}.
+   * <p>The two relational expressions are assumed as semantically equivalent,
+   * that means the hints should be attached to the relational expression
+   * that expects to have them.
+   *
+   * <p>In order to achieve this, we do the following now:
+   *
+   * <ol>
+   *   <li>If the {@code originalRel} is not Project but {@code newRel} is,
+   *   try to copy the hints to the {@code newRel}'s input, this is needed because
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/plan/RelOptUtil.java]: `is` seems redundant.

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java [hunk 4]
```diff
@@ -2495,12 +2504,6 @@ public RelBuilder permute(Mapping mapping) {
     return project(exprList);
   }
 
-  public RelBuilder aggregate(GroupKey groupKey,
-      List<AggregateCall> aggregateCalls) {
-    return aggregate(groupKey,
-        Lists.transform(aggregateCalls, AggCallImpl2::new));
-  }
-
   /** Creates a {@link Match}. */
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: Is this change expected? It seems not necessary.

File: core/src/test/java/org/apache/calcite/test/SqlHintsConverterTest.java (hunk 1 above)
Comment (amaliujia) [reply core/src/test/java/org/apache/calcite/test/SqlHintsConverterTest.java]: gotcha. Thanks for your clarification.

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java (hunk 4 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java]: Not that necessary, just move the code where it should belong.

File: core/src/main/java/org/apache/calcite/plan/RelOptUtil.java (hunk 3 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/plan/RelOptUtil.java]: which `is` ?

**Comment by danny0405** [comment]
> BTW, I noticed that the docs in `hint/package-info.java` should be updated 😄.

Yes, thanks for the reminder ~

**Pull request merged**